### ISSUE TITLE
Debug Performance scaling

### ISSUE THREAD

**Issue by okennedy** [VizierDB/vizier-scala#182]
A 100 step notebook is starting to take a while to load.  A few places to look for optimization opportunities
- [ ] Kill the HATEOAS crap in the workflow description.  It's no longer needed
- [ ] The SQLite interface seems to be a significant bottleneck.  Even with the current timeout, I'm getting actual legit timeouts (and, I think, not just reentrancy bugs).  Either this is some sort of deadlock or race condition 
- [ ] My gut feeling right now is that we're blocked on the backend... but it would be useful to confirm that we're not wasting time in the frontend (e.g., the Rx graph is likely quite large at this point).

At the same time, there's a bunch of stuff in spark that we can probably clean up
- [ ] I'm getting "Native library not found" warnings... see if this is a performance issue
- [ ] I'm getting "window query without partition" warnings... Spark claims that this is a performance issue, but is it really, at the sizes we're talking about?
- [ ] Start considering ways to cache intermediate outputs.

**Comment by okennedy** [comment]
The gut feeling was at least partly right.  There's a ton of contention on the SQLite handle, and we're (unsurprisingly) suffering from ORM-hell (a few dozen queries that could be issued as a single query)

1. Currently, a cell+module description needs to include a list of all of the artifacts visible at the module. Computing this is expensive, comes up in a ton of different places, none of which actually care about it downstream.
2. Currently, in the cell+module description, and in the workflow description, there is a distinction made between different types of artifacts. I'd also like to coalesce these together.

These two changes, together, improve GetWorkflow from 1.7s down to about 0.6s.  

A few other observations
- There's a number of places where the database lock is being held for ~2-4s.  I **think** this is some sort of queueing delay.  I'm not sure.
- Workflow.describe is still pretty pricy.  I think it might be possible to cut that down a bit by fetching the following in two bulk queries:
    - the outputRefs for each module as `ArtifactSummary`ies
    - the messages for each module


I think similar problems affect the cell update commands (although the above seem like they *should* improve the situation some).

**Comment by okennedy** [comment]
LoadDataset is the only dataset constructor that needs to go to spark for schema lookups.  Per #193, cache this information at cell-execution time and streamline the load process.

**Comment by okennedy** [comment]
- [ ] Updating a cell triggers a flood of state updates.  These could be either buffered or batched together.

**Comment by okennedy** [comment]
Still working on this, but I can report a significant improvement in large notebooks.  Reducing the number of worker and supervisor threads has actually improved performance significantly, although it seems to be busting a few of the test cases.  As far as I can tell, if the supervisor pool is exhausted, new RunningWorkflow instances block indefinitely.  Requires further investigation.

**Comment by okennedy** [comment]
We're in a far better place now than we were a month ago.  Optimization is an infinite time sink, so closing this for 2.0.  We'll likely want to revisit at some point:
- Frontend performance
- The database adaptor
- Scalikejdbc's / Scala expensive prices for hydrating tiny case classes
